### ISSUE TITLE
DIND image cache causes inode shortages

### ISSUE THREAD

**Issue by betatim** [jupyterhub/mybinder.org-deploy#371]
Around 2.30pm CET we had a [dip in launch successes](https://grafana.mybinder.org/render/dashboard-solo/db/main-dashboard?refresh=1m&orgId=1&from=1518550532196&to=1518636932197&panelId=17&width=1000&height=500&tz=UTC%2B01%3A00).

Started browsing the logs on stackdriver via the web interface. Limited it to WARNING or higher level messages. Then switched to using CLI based log reading to make it easier to reproduce.

```
gcloud logging read 'resource.type="gke_cluster" 
resource.labels.project_id="binder-prod"
resource.labels.cluster_name="prod-a"
timestamp>="2018-02-14T13:02:04.000000000Z"
severity >= WARNING
timestamp < "2018-02-14T13:02:10.000000000Z"' --format json
```

gets you roughly in the right neighbourhood of the log messages.

```
gcloud logging read 'resource.type="gke_cluster"
resource.labels.project_id="binder-prod"
resource.labels.cluster_name="prod-a"
timestamp>="2018-02-14T11:34:04.000000000Z"
severity >= WARNING
timestamp < "2018-02-14T11:37:10.000000000Z"' --format json
```
gets you what we think is kind of the start of the problems.

Conclusion with @yuvipanda:
> 1. if we don't cleanup images in dind pods, 2. they'll fill up disk at about once a month or something, 3. causing dind pod to restart / kubelet to gc images, 4. This leads to us losing both build image cache (inside dind) and run image cache (on the node)
> so I propose we prune the images inside the dind to remove everything more than 7d old. This should give us enough of a build cache without running into this

Basically we create our own version of garbage collection for images created by DIND.

Additionally taking a look at adding a panel to grafana that shows inode usage. We have a few related to disk usage but that never looked like it was anywhere close to the limit so an extra panel might be worth it.

**Comment by choldgraf** [comment]
arg I just discovered this issue! Could have saved us a bit of time later on in march :-P

Ah well - just an update on this, we now have this grafana board for inodes usage:

https://grafana.mybinder.org/dashboard/db/node-activity?refresh=1m&orgId=1&panelId=31&fullscreen&from=1525107107515&to=1525128707515

And the image cleaner is in the process of being fixed up, which should more generally solve this problem

**Comment by yuvipanda** [comment]
I don't think this happens anymore. There's also the image-cleaner now that helps.
